### PR TITLE
Intependent_Cascade

### PR DESCRIPTION
**Pull request by anty-filidor**

### CODE REVIEW

File: network_diffusion/models/ICM_model2.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+from network_diffusion.seeding.base_selector import BaseSeedSelector
+from network_diffusion.utils import BOLD_UNDERLINE, THIN_UNDERLINE, NumericType
+
+
+class ICModel(BaseModel):
+
+    INACTIVE_NODE = "0"
+    ACTIVE_NODE = "1"
+    PROCESS_NAME = "ICM"
+    DONE = "-1"
```
Comment (anty-filidor) [inline network_diffusion/models/ICM_model2.py:20]: I'd change states to:
- inactive
- active
- activated

With transitions 
inactive->active with probability p (passed in the __init__)
active->activated with probability 1 (this is an equivalent of your "done")

File: network_diffusion/models/ICM_model2.py [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        return a list of activated_nodes if the protocol allow
+
+        """
+
+        activated_nodes: List[NUBff] = []
+
+        for actor in net.get_actors():  # checking each node
+            layer_inputs = {}
+
+            if set(actor.states.values()) == set(self.ACTIVE_NODE):
```
Comment (anty-filidor) [inline network_diffusion/models/ICM_model2.py:147]: not ACTIVE but ACTIVATED!

File: network_diffusion/models/ICM_model2.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+        param_agent is node to check
+        param_layer_name is node's layer
+        param_net is choosing network
+
+        """
+        probability = self.probability
+        l_graph: nx.Graph = net.layers[layer_name]
+        current_state = l_graph.nodes[agent.actor_id]["status"]
+        if current_state == self.ACTIVE_NODE:
```
Comment (anty-filidor) [inline network_diffusion/models/ICM_model2.py:120]: then chenge state to ACTIVATED

File: network_diffusion/models/ICM_model2.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+    def __init__(
+        self,
+        seeding_budget: Tuple[NumericType, NumericType],
+        seed_selector: BaseSeedSelector,
+        protocol: str,
+        probability: float
+    ) -> None:
+        assert 0 <= probability <= 1, f"incorrect probability: {probability}!"
+        self.probability = probability
+        compart_graph = self._create_comparents(seeding_budget)
```
[the remark this answers is not in the document]
Comment (anty-filidor) [reply network_diffusion/models/ICM_model2.py:33]: If not, then please ignore it for now and try to implement model without this property

File: network_diffusion/models/ICM_model.py [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+from network_diffusion.mln.mlnetwork import MultilayerNetwork
+from network_diffusion.models.base_model import BaseModel
+from network_diffusion.mln.actor import MLNetworkActor
+from network_diffusion.models.utils.compartmental import CompartmentalGraph
+from network_diffusion.models.utils.types import NetworkUpdateBuffer as NUBff
+from network_diffusion.seeding.base_selector import BaseSeedSelector
+from network_diffusion.utils import BOLD_UNDERLINE, THIN_UNDERLINE, NumericType
+
+
+class ICModel(BaseModel):
```
Comment (anty-filidor) [inline network_diffusion/models/ICM_model.py:15]: I'd leave this class as a baseline implementation and try to focus on your second (much mode clear) approach :)

File: network_diffusion/models/ICM_model.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                    [
+                        NUBff(
+                            node_name=actor.actor_id,
+                            layer_name=layer_name,
+                            new_state=self.ACTIVE_NODE,
+                        )
+                        for layer_name in layer_inputs
+                    ]
+                )
+        return activated_nodes
```
Comment (anty-filidor) [inline network_diffusion/models/ICM_model.py:164]: you're applying protocol to actors that changed their state from active to activated, which is incorrect!

File: network_diffusion/models/mic_model.py [hunk 7]
```diff
@@ -16,75 +33,99 @@
...
     ACTIVE_NODE = "1"
-    PROCESS_NAME = "ICM"
     ACTIVATED_NODE = "-1"
-    probability = 0.0
+    PROCESS_NAME = "MICM"
 
     def __init__(
         self,
-        seeding_budget: Tuple[NumericType, NumericType],
+        seeding_budget: Tuple[NumericType, NumericType, NumericType],
```
Comment (damian4060) [inline network_diffusion/models/mic_model.py:45]: You sending budget on 3 states. It is required. It means 2 states are not enough. Maybe you take in account the next epochs when this state is

Comment (anty-filidor) [reply network_diffusion/models/mic_model.py:45]: I comprehend a seeding budget as information about percentages of agents in each state defined for the process. For instance, we can have: [32,68,0] -> 32% are inactive, 68% are active, and 0% are activated before the simulation

**Review by pbrodka — APPROVED** [review]

**Pull request merged**